src/errorhandling/SDKErrors.ts

```diff
@@ -0,0 +1,493 @@
+/**
+ * SDKErrors are KILT-specific errors, with associated codes and descriptions.
+ *
+ * @packageDocumentation
+ * @module SDKError
+ * @preferred
+ */
+
+import { NonceHash } from '../types/RequestForAttestation'
+
+export enum ErrorCode {
+  ERROR_CTYPE_HASH_NOT_PROVIDED = 10001,
+  ERROR_CLAIM_HASH_NOT_PROVIDED = 10002,
+  ERROR_CLAIM_HASHTREE_NOT_PROVIDED = 10003,
+  ERROR_CLAIM_NOT_PROVIDED = 10004,
+  ERROR_OWNER_NOT_PROVIDED = 10005,
+  ERROR_RFA_NOT_PROVIDED = 10006,
+  ERROR_ATTESTATION_NOT_PROVIDED = 10007,
+  ERROR_REVOCATION_BIT_MISSING = 10008,
+  ERROR_LEGITIMATIONS_NOT_PROVIDED = 10009,
+  ERROR_ATTESTATION_SESSION_MISSING = 10010,
+  ERROR_PE_MISSING = 10011,
+  ERROR_PE_CREDENTIAL_MISSING = 10012,
+
+  ERROR_ADDRESS_TYPE = 20001,
+  ERROR_HASH_TYPE = 20002,
+  ERROR_HASH_MALFORMED = 20003,
+  ERROR_NONCE_HASH_TYPE = 20004,
+  ERROR_SIGNATURE_DATA_TYPE = 20005,
+  ERROR_CTYPE_OWNER_TYPE = 20006,
+  ERROR_DELEGATION_ID_TYPE = 20007,
+  ERROR_NONCE_HASH_MALFORMED = 20008,
+  ERROR_CLAIM_CONTENTS_MALFORMED = 20009,
+  ERROR_CLAIM_HASHTREE_MALFORMED = 20010,
+  ERROR_OBJECT_MALFORMED = 20011,
+  ERROR_MNEMONIC_PHRASE_MALFORMED = 20012,
+  ERROR_QUOTE_MALFORMED = 20013,
+  ERROR_CLAIM_HASHTREE_MISMATCH = 20014,
+  ERROR_PE_MISMATCH = 20015,
+  ERROR_DID_IDENTIFIER_MISMATCH = 20016,
+  ERROR_ROOT_NODE_QUERY = 20017,
+  ERROR_INVALID_DID_PREFIX = 20018,
+
+  ERROR_ADDRESS_INVALID = 30001,
+  ERROR_NONCE_HASH_INVALID = 30002,
+  ERROR_LEGITIMATIONS_UNVERIFIABLE = 30003,
+  ERROR_SIGNATURE_UNVERIFIABLE = 30004,
+  ERROR_ATTESTEDCLAIM_UNVERIFIABLE = 30005,
+  ERROR_CLAIM_UNVERIFIABLE = 30006,
+  ERROR_CTYPE_HASH_INVALID = 30007,
+  ERROR_MNEMONIC_PHRASE_INVALID = 30008,
+  ERROR_IDENTITY_MISMATCH = 30009,
+  ERROR_ROOT_HASH_UNVERIFIABLE = 30010,
+
+  ERROR_DECOMPRESSION_ARRAY = 40001,
+  ERROR_COMPRESS_OBJECT = 40002,
+  ERROR_DECODING_MESSAGE = 40003,
+  ERROR_PARSING_MESSAGE = 40004,
+  ERROR_MESSAGE_TYPE = 40005,
+
+  ERROR_UNKNOWN = -1,
+}
+
+export class SDKError extends Error {
+  public errorCode: ErrorCode
+
+  public constructor(errorCode: ErrorCode, message: string) {
+    super(message)
+    this.errorCode = errorCode
+  }
+}
+export const ERROR_CTYPE_HASH_NOT_PROVIDED: () => SDKError = () => {
+  return new SDKError(
+    ErrorCode.ERROR_CTYPE_HASH_NOT_PROVIDED,
+    'CType hash missing'
+  )
+}
```

Comment:
@wischli Hey I just noticed that it is a bit awkward  to check for the specific error. Would it be better to use Classes?

```suggestion
export class CTypeHashNotProvidedError extends SDKError {
  public constructor() {
    super(
      ErrorCode.ERROR_CTYPE_HASH_NOT_PROVIDED,
      'CType hash missing'
    )
  }
}
```

If you concur I could do the refactoring. 😋

Reply:
I am not sure whether to agree. I like the current namespaces of the SDK errors as they follow the format of the ExtrinsicErrors. Have you tried the following?
```javascripts
expect(() => thingThatFails()).toThrowError(ERROR_CTYPE_HASH_NOT_PROVIDED())
```

Reply:
Ah yes I see. That makes sense. It's the same kind of API as for the ExtrinsicsErrors. But what about:

```
expect(() => AttestationUtils.errorCheck(malformedClaimHash)).toThrowError(
  SDKErrors.ERROR_HASH_MALFORMED(malformedClaimHash.claimHash, 'Claim')
)
```

I think it is a bit hard to know for a developer what we would pass inside the SDK. But I understand your argument. If you prefer  keeping it consistent, that would be fine for me.

Reply:
I am unsure but I suppose this should be handled in another ticket if we decide on refactoring. Since Leon just merged this, I would continue as is unless we run in lots of boilerplate or trouble.

Reply:
Hö? Ok no then it's off topic! I thought this was added in this PR... It was shown in the diff... 😕

Reply:
Yeah I don't know why it's shown in the diff. It's a PR Leon has worked on, reviewed by me 😅